File: instrumentation/java-http-server/library/src/main/java/io/opentelemetry/instrumentation/javahttpserver/internal/Experimental.java
/*
 * Copyright The OpenTelemetry Authors
 * SPDX-License-Identifier: Apache-2.0
 */

package io.opentelemetry.instrumentation.javahttpserver.internal;

import io.opentelemetry.instrumentation.javahttpserver.JavaHttpServerTelemetryBuilder;
import java.util.function.BiConsumer;
import javax.annotation.Nullable;

/**
 * This class is internal and experimental. Its APIs are unstable and can change at any time. Its
 * APIs (or a version of them) may be promoted to the public stable API in the future, but no
 * guarantees are made.
 */
public final class Experimental {

  @Nullable
  private static volatile BiConsumer<JavaHttpServerTelemetryBuilder, Boolean>
      setEmitExperimentalTelemetry;

  public static void setEmitExperimentalTelemetry(
      JavaHttpServerTelemetryBuilder builder, boolean emitExperimentalTelemetry) {
    if (setEmitExperimentalTelemetry != null) {
      setEmitExperimentalTelemetry.accept(builder, emitExperimentalTelemetry);
    }
  }

  public static void internalSetEmitExperimentalTelemetry(
      BiConsumer<JavaHttpServerTelemetryBuilder, Boolean> setEmitExperimentalTelemetry) {
    Experimental.setEmitExperimentalTelemetry = setEmitExperimentalTelemetry;
  }

  private Experimental() {}
}

File: instrumentation/ktor/ktor-2-common/library/src/main/kotlin/io/opentelemetry/instrumentation/ktor/v2_0/common/client/internal/Experimental.kt
/*
 * Copyright The OpenTelemetry Authors
 * SPDX-License-Identifier: Apache-2.0
 */

package io.opentelemetry.instrumentation.ktor.v2_0.common.client.internal

import io.opentelemetry.instrumentation.ktor.v2_0.common.client.AbstractKtorClientTelemetryBuilder
import java.util.function.BiConsumer

class Experimental private constructor() {

  companion object {
    @Volatile
    private var setEmitExperimentalTelemetry: BiConsumer<AbstractKtorClientTelemetryBuilder, Boolean>? = null

    fun emitExperimentalTelemetry(builder: AbstractKtorClientTelemetryBuilder) {
      if (setEmitExperimentalTelemetry != null) {
        setEmitExperimentalTelemetry!!.accept(builder, true)
      }
    }

    fun setSetEmitExperimentalTelemetry(setEmitExperimentalTelemetry: BiConsumer<AbstractKtorClientTelemetryBuilder, Boolean>?) {
      Companion.setEmitExperimentalTelemetry = setEmitExperimentalTelemetry
    }
  }
}
